Set User.UID from x509 cert

modulitos opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

I will followup this PR with 2 more changes:
 - update kubeadm to include UID in the cert
 - update the docs, [per this comment](https://github.com/kubernetes/kubernetes/pull/127897#discussion_r1797458976)

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:
This PR allows users to optionally set the `User.UID` from an x509 client cert, instead of always having it empty.  This is part of an implementation to resolve https://github.com/kubernetes/kubernetes/issues/127860

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Addresses https://github.com/kubernetes/kubernetes/issues/127860

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubernetes components that accept x509 client certificate authentication now read the user UID from a certificate subject name RDN with object id 1.3.6.1.4.1.57683.2. An RDN with this object id must contain a string value, and appear no more than once in the certificate subject. Reading the user UID from this RDN can be disabled by setting the beta feature gate `AllowParsingUserUIDFromCertAuth` to false (until the feature gate graduates to GA).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
N/A

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

modulitos committed:
set user.DefaultInfo.UID from x509 cert

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:
```diff
@@ -37,9 +37,31 @@ const (
+	   S+IMBArINm7w+VJUqbyg5tJTv0Sx4BUJWkPGmkb0uq9y2ntCO21CARECQCT0BMIL
+	   nQ+uHPh8Y7nTvlJ5VGU/eW1skd9y4/9wvJz3/YAD0H7/r7PgjtYoozpHpjDUbcnm
+	   1qPcKn6wFSZwlEA=
+	   -----END PRIVATE KEY-----
```
Documenting the `root.key` here, so that it can be re-used. This avoid having to regenerate the root cert when making changes (eg: adding a `UID` field to the client cert).

enj replied:
I would prefer to just re-generate to avoid having a test private key embedded here.  This is a slow moving package.

modulitos replied:
Done 👍

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:
```diff
@@ -277,63 +299,64 @@ u7ESEaSlma8XEeex
-        Subject: C = US, ST = My State, L = My City, O = My Org, OU = My Unit, CN = client_cn
+            Not Before: Oct  6 20:50:43 2024 GMT
+            Not After : Sep 12 20:50:43 2124 GMT
+        Subject: C=US, ST=My State, L=My City, O=My Org, OU=My Unit, CN=client_cn, UID=test123
```
UID added to the client cert here

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:
```diff
@@ -640,8 +663,9 @@ func TestX509(t *testing.T) {
 				User: &user.DefaultInfo{
 					Name:   "client_cn",
 					Groups: []string{"My Org"},
+					UID:    "test123",
```
This is the behavioral change to assert `User.UID` matches that of the cert.

micahhausler commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:
```diff
@@ -37,9 +37,31 @@ const (
+
+	   > output
+
+	   -----BEGIN PRIVATE KEY-----
```
Its obvious to us, but put a note here that this is for testing only. K8s gets all kinds of hacker reports for test keys and they seem to more often ignore if there's a comment immediately preceding the key

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +285,28 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+// we read it off the asn1 array.
+func parseUIDFromCert(cert *x509.Certificate) string {
+	for _, name := range cert.Subject.Names {
+		if name.Type.Equal(uidOID) {
```
I assume this `if` statement can only evaluate to `true` for one `name`?  i.e. can a cert have more than one `UID` set?

enj replied:
> [2.39](https://datatracker.ietf.org/doc/html/rfc4519#section-2.39).  'uid'
> 
>    The 'uid' ('userid' in [RFC 1274](https://datatracker.ietf.org/doc/html/rfc1274)) attribute type contains computer
   system login names associated with the object.  Each name is one
   value of this multi-valued attribute.
   (Source: [RFC 2798](https://datatracker.ietf.org/doc/html/rfc2798) [[RFC2798](https://datatracker.ietf.org/doc/html/rfc2798)] and [RFC 1274](https://datatracker.ietf.org/doc/html/rfc1274) [[RFC1274](https://datatracker.ietf.org/doc/html/rfc1274)])

sftim replied:
The OID that applies here is 0.9.2342.19200300.100.1.1 and I think (not sure!) that one is single valued.

modulitos replied:
Very interesting case - thanks everyone for the feedback!

I pushed an update to handle multi-valued `UID`s by selecting the last `UID`. This is consistent with how the `x509/pkix` package handles multi-valued `CN` attribute types: https://cs.opensource.google/go/go/+/master:src/crypto/x509/pkix/pkix.go;l=161?q=FillFromRDNSequence&ss=go%2Fgo
I updated the test to verify behavior when there are multiple `CN` and `UID` attribute types.

Note how `CN` is a `multi-valued` attribute as well:

>   The 'cn' ('commonName' in X.500) attribute type contains names of an
>   object.  Each name is one value of this multi-valued attribute.  If
>   the object corresponds to a person, it is typically the person's full
>   name.

https://datatracker.ietf.org/doc/html/rfc4519#section-2.3

Also note that the order of multi-valued attribute types follow the serialized representation in the ASN.1 structure of X.509 certificate:   https://cs.opensource.google/go/go/+/master:src/crypto/x509/x509.go;l=2142-2149?q=FillFromRDNSequence&ss=go%2Fgo
(ie: they are ordered according to how they were passed into the cert upon creation - there is no additional sorting being done)

liggitt replied:
I'm not sure the `uid` defined in that RFC matches what we mean by user uid

What we mean (the "u" stands for "unique"):
```
	// GetUID returns a unique value for a particular user that will change
	// if the user is removed from the system and another user is added with
	// the same name.
	GetUID() string
```

RFC (the "u" stands for "user"):
```
The 'uid' ('userid' in [RFC 1274](https://datatracker.ietf.org/doc/html/rfc1274)) attribute type contains computer system login names associated with the object.  Each name is one
```

If we were going to use a pre-defined x509 value, `x500UniqueIdentifier` seems much closer to what we mean by `uid`:

https://datatracker.ietf.org/doc/html/rfc4519#section-2.43

> The 'x500UniqueIdentifier' attribute type contains binary strings that are used to distinguish between objects when a distinguished name has been reused.  Each string is one value of this multi-valued attribute.

That said, that seems problematic because it's binary, not text, and it's still multi-valued, and it seems very strange to intentionally map a multi-valued attribute to a single-value in our user interface, and accept certs that specify multiple values for it (if we'd known CN was multi-value, we likely would not have used it for username).

liggitt replied:
I think it's _probably_ safer to obtain / define an oid exactly for our use, to eliminate the chance of any well-behaved CA already issuing certs using a "standard" oid in ways that will cause us issues (putting non-utf8 data in a binary string field, or actually putting multiple values in a multi-value attribute, etc)

sftim replied:
If we're assigning OIDs, we could perhaps annotate certificates to say "look, for this particular certificate, treat RFC 1274 `userid` as the username, get groups from this multivalue username, and ignore commonName completely".

I'd support a PR that did that other more complex thing. But: complex, it is.

liggitt replied:
I really want to keep backwards compat with existing username / group extraction for n-1 API servers and n-1/2/3 kubelets

sftim replied:
I was thinking _n+(many)_.

Maybe phase out the existing thing after 10 minor releases? Maybe leave it in even longer behind a flag?
I wish it didn't need a KEP, but it would.

Anyway, I mention the more complex option in case we pick something that makes moving away from `commonName` hard, given we may one day want to.

enj replied:
> A new attribute for extracting uid is a strictly additive change we don't have to worry about breaking old servers / kubelets if they encounter a client certificate which contains it

+1 to this approach.  If we want to start messing with `CN`, then that is whole another conversation that will 100% require a KEP and out of scope for this PR.

modulitos replied:
I see a couple approaches for defining a new OID, but not sure how to proceed:
1. It seems [OID 1.3.9900 to 1.3.9999 may be considered as "internal use" OIDs](http://oid-info.com/get/1.3). I pushed an implementation using `1.3.9901.1` as the OID - see the latest commit exemplifying this.
2. We can [apply for a private enterprise number through IANA here](https://www.iana.org/assignments/enterprise-numbers/assignment/apply/). Registration is free, and it takes ~1 week. But who should file it, or which name/address should I use? Happy to help where I can.

(I also posted this question in our related [slack discussions with more details](https://kubernetes.slack.com/archives/C05MW3382US/p1729976457287509?thread_ts=1729791759.941219&cid=C05MW3382US))
Thank you for the feedback :)

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +285,28 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+func parseUIDFromCert(cert *x509.Certificate) string {
+	for _, name := range cert.Subject.Names {
+		if name.Type.Equal(uidOID) {
+			if s, ok := name.Value.(string); ok {
```
```suggestion
			if s, ok := name.Value.(string); ok && len(s) > 0 {
```

sftim commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +285,28 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
 	return &authenticator.Response{
 		User: &user.DefaultInfo{
 			Name:   chain[0].Subject.CommonName,
```
IMO, either:

- We should update the documentation on authentication to make it clear that we always use CN as username, and never UID.
- We should support certificates with userid set but no common name, and in that case we popular both UID and User from userid.

Unfortunately, we use Organization and CommonName in a somewhat off-label formulation, but I also think we'll have to support that approach for many years to come.

modulitos replied:
Good points! I'm open to either approach, and I'm willing to defer to other decision makers.

That being said, I think the first option is preferable because it's simpler: CN sets the `user.Name`, and UID sets the `user.UID`, without any conditionals.

The second option is appealing, but I'm not sure there's a need to fallback (ie: setting `user.Name` from the UID if CN is empty). We can always add it as a followup if a use case arises.

Assuming we proceed with the first option, I'm happy to update the docs once this PR is approved. I found these sections where I can clarify the CN and UID expectations:

https://kubernetes.io/docs/reference/access-authn-authz/authentication/#users-in-kubernetes
> Kubernetes determines the username from the common name field in the 'subject' of the cert (e.g., "/CN=bob"). 

and "X509 client certificates" here:
https://kubernetes.io/docs/reference/access-authn-authz/authentication/#x509-client-certificates

dims commented:
enough eyes to try running CI jobs

/ok-to-test

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509_test.go:
```diff
@@ -640,8 +642,9 @@ func TestX509(t *testing.T) {
+					UID:    "client_id",
 					Extra: map[string][]string{
-						user.CredentialIDKey: {"X509SHA256=dd0a6a295055fa94455c522b0d54ef0499186f454a7cf978b8b346dc35b254f7"},
+						"authentication.kubernetes.io/credential-id": {"X509SHA256=4f830ed74c460c0b14b177e02ee370a8bd38c61c7b813c607255245a5e0b3b4e"},
```
```suggestion
						user.CredentialIDKey: {"X509SHA256=4f830ed74c460c0b14b177e02ee370a8bd38c61c7b813c607255245a5e0b3b4e"},
```

enj commented:
Per the [slack thread](https://kubernetes.slack.com/archives/C05MW3382US/p1729869014699899?thread_ts=1729791759.941219&cid=C05MW3382US), lets gate this behind a beta, on-by-default feature gate.

neolit123 commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +285,28 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+
+// Using CNCF PEN for Kubernetes:
+// https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/4317-pod-certificates#x509-podidentity-extension
+var uidOID = asn1.ObjectIdentifier([]int{1, 3, 6, 1, 4, 1, 57683, 2})
```
if you want other certificate signing components to re-use this it should go in e.g. k8s.io/component-base; k8s.io/apiserver doesn't seem that appropriate

liggitt replied:
All the existing x509 client validation lives in staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go currently

This should go with that; if we intend to relocate x509 authentication, all of it should relocate together

enj replied:
If other parts of the codebase want to create a client cert, they only need to rely on the std lib because of the current CN and O usage.  With this change you need to know this magic OID, so I could see some argument for putting it in k8s.io/api or similar.

modulitos replied:
Thanks for pointing this out.

Performing the refactor in this PR seems a bit much, but I'll explore doing that next. I'll have a followup PR to use the new OID in kubeadm, which is a good opportunity to think about this re-usability issue.

neolit123 commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +285,28 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+
+// Go's crypto/x509 library doesn't explicitly serialize UID to a named field, so
+// we read it off the asn1 array.
+func parseUIDFromCert(cert *x509.Certificate) string {
```
same for this

modulitos replied:
Addressed in this discussion: https://github.com/kubernetes/kubernetes/pull/127897#discussion_r1825031398

fedebongio commented:
/remove-sig api-machinery

cici37 commented:
/remove-sig api-machinery

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -281,13 +284,36 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
 })
+
+// Using CNCF PEN for Kubernetes:
+// https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/4317-pod-certificates#x509-podidentity-extension
```
```suggestion
// https://www.iana.org/assignments/enterprise-numbers.txt
//
// 57683
//   Cloud Native Computing Foundation
//     Davanum Srinivas
//       sc1&kubernetes.io
```

modulitos replied:
Done 👍

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -281,13 +284,36 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+// https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/4317-pod-certificates#x509-podidentity-extension
+var uidOID = asn1.ObjectIdentifier([]int{1, 3, 6, 1, 4, 1, 57683, 2})
+
+// Go's crypto/x509 library doesn't explicitly serialize UID to a named field, so
+// we read it off the asn1 array.
```
This comment doesn't really make sense when using a PEN.

modulitos replied:
Thanks for the catch - I removed it 👍

enj reviewed (changes requested):
Minor comments.  @liggitt PTAL for API review.

@kubernetes/steering-committee any concerns with this PEN usage?  Looks like @dims was the one the registered it.

dims commented:
> @kubernetes/steering-committee any concerns with this PEN usage? Looks like @dims was the one the registered it.

context: https://github.com/kubernetes/k8s.io/issues/1959

modulitos commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -263,6 +268,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
+	AllowParsingUserUIDFromCertAuth: {
+		{Version: version.MustParse("1.32"), Default: true, PreRelease: featuregate.Beta},
+	},
+
```
Done 👍

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -281,13 +284,36 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+
+// Go's crypto/x509 library doesn't explicitly serialize UID to a named field, so
+// we read it off the asn1 array.
+func parseUIDFromCert(cert *x509.Certificate) string {
```
Done 👍

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -281,13 +284,36 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
 	return &authenticator.Response{
 		User: &user.DefaultInfo{
 			Name:   chain[0].Subject.CommonName,
+			UID:    uid,
```
Done 👍

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -281,13 +284,36 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+		if name.Type.Equal(uidOID) {
+			if s, ok := name.Value.(string); ok && len(s) > 0 {
+				return s
+			}
```
Good idea. I made these changes and added a unit test for when multiple UIDs are set.

For context, I was initially trying to keep consistent with how we parse User.Name from the CN, but that doesn't seem desirable [per an earlier discussion](https://github.com/kubernetes/kubernetes/pull/127897#discussion_r1790460482). Would be nice to fix that, but it's a separate effort.

enj commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +287,45 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+			return ""
+		}
+		found = true
+		if s, ok := name.Value.(string); ok && len(s) > 0 {
+			uid = s
+		}
```
nit:

```suggestion
		uid, _ = name.Value.(string)
```

enj reviewed (commented):
Minor comment, this LGTM pending API review.

enj commented:
/kind api-change

enj commented:
/milestone v1.32

fedebongio commented:
/remove-sig api-machinery
/triage accepted

modulitos commented:
I've done the [API review checklist](https://github.com/kubernetes/community/blob/master/sig-architecture/api-review-process.md#mechanics), please review :)

Note that KEP wasn't considered necessary, per this thread in `#sig-auth`: https://kubernetes.slack.com/archives/C0EN96KUY/p1727984474325949

Here is another thread with context in `#sig-auth-certificates-dev`: https://kubernetes.slack.com/archives/C05MW3382US/p1729791759941219

knabben commented:
Hey @modulitos @enj @micahhausler 
⚠️ Do we still intend to merge this for v1.32? Just a reminder that the code freeze is starting 02:00 UTC Friday November 8th 2024 (a little less than 1 week from now). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

modulitos commented:
@knabben Yes, we intend to merge this for 1.32. It's not urgent, but would be nice to get it in.

We are waiting on api review - perhaps @liggitt can take a look 🤞

liggitt commented:
> We are waiting on api review - perhaps @liggitt can take a look 🤞

it's in my list but the last two weeks have been nuts with reviews and rolling back regressions :-/

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +287,43 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+		}
+		if found {
+			// Require that our OID is only used once.
+			// If there are multiple, we leave UID unset.
```
Done 👍

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +287,43 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+			return ""
+		}
+		found = true
+		uid, _ = name.Value.(string)
```
Updated this to return an error if string parsing isn't successful 👍 

I was assuming that strings were always parsed as such, but I'll need to verify. I'll dig into the asn1 parsing, and whether we can surface a non-string in a test 🤔

modulitos replied:
> is there a way to exercise the non-string case in a test?

Yes, this is possible, but I'm not sure if it's worth the effort. Here's a commit that tests the behavior: https://github.com/kubernetes/kubernetes/pull/127897/commits/ef3775610ff70aa23b056d1d03e1f2da43c805c4

but I rolled it back because I don't think it's the proper way to "edit" a cert. Instead, we'd need to generate a new cert after setting the binary field.

> I'm fuzzy on asn1 parsing, do strings always show up as strings or can they be []byte depending on the encoder?

Great question. Yes, I believe it's possible parse a cert with a []byte as a field. I'd love to prove this in a test, now that we've added the error handling. I'm also willing to do it as a followup, if we want to merge this PR sooner.

modulitos replied:
After more investigation, I don't think it's possible to generate a cert with a non-string value.

Some details are in [my latest comment](https://github.com/kubernetes/kubernetes/pull/127897#issuecomment-2467209437)

Sorry for the back-and-forth!

modulitos commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -284,10 +287,43 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+//	    sc1&kubernetes.io
+var uidOID = asn1.ObjectIdentifier([]int{1, 3, 6, 1, 4, 1, 57683, 2})
+
+func parseUIDFromCert(cert *x509.Certificate) string {
```
Done 👍

modulitos commented:
@liggitt  Thanks for the feedback 🙌 

Requested changes have been implemented, although I'm still trying to surface a test case where the UID fails to parse as a string. I'm willing to finish that in a followup PR, as it may take some time for me to grok asn1 encoding.

BenTheElder commented:
First I've heard of PEN, I guess from https://github.com/kubernetes/k8s.io/issues/1959 and skimming some more context we can delegate this to infra/auth.

I left a comment on that thread about the contact info.

I haven't had time to familiarize myself with this but if the ID is "CNCF" we should probably be talking to them about it and considering what happens when another CNCF project needs to use this? Has anyone done that yet @dims?

cc @upodroid @ameukam

dims commented:
@BenTheElder So far no other projects have approached me directly by my name or when i had sc1@ access. 

Here are the usages:
- https://github.com/kubernetes/k8s.io/issues/1959 (original request)
- https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/4317-pod-certificates/README.md#:~:text=1%20OID%20(currently-,1.3.6.1.4.1.57683.1,-%2C%20though%20we%20need (KEP-4317)

The same form can be used to modify the email id i believe if Steering wants to hand it off to CNCF.

liggitt commented on staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go:
```diff
@@ -281,13 +284,45 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
+		if !ok {
+			return "", fmt.Errorf("unable to parse UID into a string")
+		}
+		uids = append(uids, uid)
```
@enj do you think we should require the uid be non-empty if specified in a subject name, or tolerate it being explicitly present and set to an empty string?

modulitos replied:
I updated UID parsing to return an error for the empty string case. I don't think an empty string would every be desirable, so returning an error seems like the best user experience.

Note that I didn't add a test because it doesn't seem possible to pass in empty values using the `openssl` command. For example, calling `openssl` with `-subj "/kube_uid="` elides the oid from `Certificate.Subject.Names`. (whereas `openssl` with `-subj "/kube_uid=abc"` includes it)

liggitt commented:
I'm happy with the shape of this, just the following questions outstanding:
* empty string handling
* whether asn1 parsing always goes to a string
* whether to test the non-string case

You can also squash to a single commit at this point

liggitt commented:
@tallclair it looks like all the InPlacePodVerticalScaling tests are failing in the alpha job... is that expected?

liggitt commented:
(dropping from the milestone until it's decided whether this will request / get an exception)

modulitos commented:
> dropping from the milestone until it's decided whether this will request / get an exception

Either way sounds good to me. Lmk if I should bump the feature gate from 1.32 to 1.33.

As for the outstanding questions:

> empty string handling

I added error handling for an empty string, just in case, to ensure the best user experience. Added some details about it in [the related thread](https://github.com/kubernetes/kubernetes/pull/127897#discussion_r1834559656).

> whether asn1 parsing always goes to a string

Yes, I'm nearly certain that asn1 parsing always goes to a string.

We can see how [parseASN1String always returns a string here](https://cs.opensource.google/go/go/+/master:src/crypto/x509/parser.go;l=60-108).

(and `parseASN1String` is called by [parseName](https://cs.opensource.google/go/go/+/master:src/crypto/x509/parser.go;l=139) which is called by [parseCertificate](https://cs.opensource.google/go/go/+/master:src/crypto/x509/parser.go;l=926))

Here's a rough overview of the code I used to test cert creation with a binary subject attribute: https://go.dev/play/p/bh_E7BMkoXp

> whether to test the non-string case

Given that asn1 parsing always goes to a string, I don't think we should add a unit test for the non-string case.
But I'll keep the code to raise an error for the non-string use case, just in case we miss something or the x509 library changes.

modulitos commented:
Seems like this missed the 1.32 release train, so I bumped this to the 1.33 release: https://github.com/kubernetes/kubernetes/pull/127897/commits/7132b274a9a21f3d83d051d5b41f582a690570d4

I also added a couple tests for the "empty string" an "non-string" error cases: https://github.com/kubernetes/kubernetes/pull/127897/commits/5b5d08e99ad4bf965d34224ef7c78ee2c86f154f

liggitt reviewed (commented):
/lgtm
/approve
/milestone v1.33

Jefftree commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -273,6 +278,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
+	AllowParsingUserUIDFromCertAuth: {
+		{Version: version.MustParse("1.33"), Default: true, PreRelease: featuregate.Beta},
```
hmm I'm a bit surprised this was able to merge without CI issues considering we haven't bumped the default binary version. cc @aaron-prindle @siyuanfoundation

siyuanfoundation replied:
It used `featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, feature.DefaultFeatureGate, version.MustParse("1.33"))` to make the unit test work.
